### ISSUE TITLE
[Bug] [Messaging] Android OnMessageReceived not being called when app is in background only

### ISSUE THREAD

**Issue by MartinGonzalez** [firebase/quickstart-unity#1020]
* Unity editor version: 2019.4.16
  * Firebase Unity SDK version: 7.0.2
  * Source you installed the SDK: Unity Package Manager
  * Problematic Firebase Component: Messaging
  * Other Firebase Components in use: Crashltycs, Analytics, Dynamic Links
  * Additional SDKs you are using: Facebook, Appsflyer, Braze
  * Platform you are using the Unity editor on: Mac
  * Platform you are targeting: iOS, Android
  * Scripting Runtime: IL2CPP

The problem is happening only on Android 8 or greater, when the app is in background only. (When app is dead is working correctly)

In Android, seems to be greater than Android 7, notification arrives correctly and when tapping it opens the app but event OnMessageReceived is not being executed.

I've tested on two devices, one with Android 7 and other Android 9, in Android 7 works everything fine but in Android 9 happens what I described above. I've read the from Android 8 or greater there is some king of priority when applications are in background but we could not found some fix to this.

We have our own Activity so we added the following code that is being called correctly and working on Android 7

```java
@Override
    public void onNewIntent(Intent intent) {
        Log.d(LogTag, "Firebase on new Intent");
        Activity activity = UnityPlayer.currentActivity;
        Intent message = new Intent(activity, MessageForwardingService.class);
        message.setAction(MessageForwardingService.ACTION_REMOTE_INTENT);
        message.putExtras(intent);
        message.setData(intent.getData());
        activity.startService(message);
    }
```

**Comment by paulinon** [comment]
Hi @MartinGonzalez,

It seems that the scenario you described and the code you provided related to Android. Is your app built using Unity? If so, I'd like you to try our quickstart (with the latest version of the SDK) and see if the issue persists.

Thanks.

**Comment by MartinGonzalez** [comment]
> Hi @MartinGonzalez,
> 
> It seems that the scenario you described and the code you provided related to Android. Is your app built using Unity? If so, I'd like you to try our quickstart (with the latest version of the SDK) and see if the issue persists.
> 
> Thanks.

Yes, it's built in with Unity but we use our own UnityPlayerActivity, we added the code that documentation says and in Android 7 is working as expected, the problem is in Android 8 or greater. 

I will try the quickstart and bring more info

**Comment by MartinGonzalez** [comment]
Hi @paulinon 

We've tested the quickstart example and also is not working, we followed the documentation and the result is the same as we have in our product, seems to be a Firebase bug or an Android limitation.

We are using MessagingUnityPlayerActivity in the quickstart AndroidManifest.xml

```xml
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android" xmlns:tools="http://schemas.android.com/tools" package="com.google.FirebaseUnityMessagingTestApp.dev" android:versionCode="1" android:versionName="1.0">
  <application android:label="@string/app_name" android:icon="@drawable/app_icon">
    <!-- The MessagingUnityPlayerActivity is a class that extends
         UnityPlayerActivity to work around a known issue when receiving
         notification data payloads in the background. -->
    <activity android:name="com.google.firebase.MessagingUnityPlayerActivity" android:configChanges="fontScale|keyboard|keyboardHidden|locale|mnc|mcc|navigation|orientation|screenLayout|screenSize|smallestScreenSize|uiMode|touchscreen">
      <intent-filter>
        <action android:name="android.intent.action.MAIN" />
        <category android:name="android.intent.category.LAUNCHER" />
      </intent-filter>
      <meta-data android:name="unityplayer.UnityActivity" android:value="true" />
    </activity>
    <service android:name="com.google.firebase.messaging.MessageForwardingService" android:exported="false" />
  </application>
</manifest>
``` 

We perform the following tests:

**In Android 8**
First Use case
* We opened the app, registered the Push token and we **killed the app.**
* We went to the firebase console created a new notification and added Data with `key` "test" and `value` "notification1" and sent it. The notification arrived correctly, we tapped on it and opened the app showing that data.
![image](https://user-images.githubusercontent.com/4130008/114233356-8bb28580-9953-11eb-9ed6-f0637c775bd1.png)

Second Use Case
* We opened the app, registered the Push token and we sent the app to background.
* We went to the firebase console created a new notification and added Data with `key` "test" and `value` "notification2" and sent it. The notification arrived correctly, we tapped on it and opened the app **did not show the data**.
![image](https://user-images.githubusercontent.com/4130008/114233496-cd433080-9953-11eb-9e2f-521f077ff613.png)

**In Android 7**
We've tested with the same `apk` the Second Use case and the data was shown correctly
![image](https://user-images.githubusercontent.com/4130008/114233566-e9df6880-9953-11eb-8991-69acd910c161.png)


Here is the [project sample](https://drive.google.com/file/d/1jGrpJFnDiYU0XQkuUMaz16mYqTEc5wdz/view?usp=sharing)

# Conclusion

* Seems to be a Firebase Bug or Android Limitation since we did not used any custom code.

**Comment by paulinon** [comment]
Thanks for providing additional details, @MartinGonzalez. I have filed an internal bug report for this. I'll be leaving this open so that any updates about this can be provided.

**Comment by vimanyu** [comment]
Hi,
This might be related to https://github.com/firebase/quickstart-unity/issues/877.
The next Unity release should contain the fix mentioned in that issue but until then, you could try the manual workaround mentioned in this comment,
https://github.com/firebase/quickstart-unity/issues/877#issuecomment-796305806
